grails-app/controllers/au/org/ala/specieslist/WebServiceController.groovy

```diff
@@ -324,4 +327,96 @@ class WebServiceController {
             render status:400, text: 'Required param not provided: email'
         }
     }
+
+    /**
+     * Lists all unique keys from the key value pairs of all records owned by the requested data resource(s)
+     *
+     * @param druid one or more DR UIDs (comma-separated if there are more than 1)
+     * @return JSON list of unique key names
+     */
+    def listKeys() {
+        if (!params.druid) {
+            response.status = HttpStatus.SC_BAD_REQUEST
+            response.sendError(HttpStatus.SC_BAD_REQUEST, "Must provide a comma-separated list of druid value(s).")
+        } else {
+            List<String> druids = params.druid.split(",")
+
+            def kvps = SpeciesListKVP.withCriteria {
+                'in'("dataResourceUid", druids)
+
+                projections {
+                    distinct("key")
+                }
+
+                order("key")
+            }
+
+            render kvps as JSON
+        }
+    }
+
+    /**
+     * Finds species list items that contains specific keys
+     *
+     * @param druid one or more DR UIDs (comma-separated if there are more than 1). Mandatory.
+     * @param keys one or more KVP keys (comma-separated if there are more than 1). Mandatory.
+     * @param format either 'json' or 'csv'. Optional - defaults to json. Controls the output format.
+     *
+     * @return if format = json, {scientificName1: [{key: key1, value: value1}, ...], scientificName2: [{key: key1, value: value1}, ...]}. If format = csv, returns a CSV download with columns [ScientificName,Key1,Key2...].
+     */
+    def listItemsByKeys() {
+        if (!params.druid || !params.keys) {
+            response.status = HttpStatus.SC_BAD_REQUEST
+            response.sendError(HttpStatus.SC_BAD_REQUEST, "Must provide a comma-separated list of druid value(s) and a comma-separated list of key(s). Parameter 'format' is optional ('json' or 'csv').")
+        } else {
+            List<String> druids = params.druid.split(",")
+            List<String> keys = params.keys.split(",")
+
+            def listItems = SpeciesListItem.withCriteria {
+                'in'("dataResourceUid", druids)
+                kvpValues {
+                    'in'("key", keys)
+                }
+
+                order("rawScientificName")
+            }
+
+            Map<String, List> results = [:]
+
+            listItems.each {
+                if (!results[it.rawScientificName]) {
+                    results[it.rawScientificName] = []
+                }
+                List kvps = []
+                it.kvpValues.each {
+                    if (keys.contains(it.key)) {
+                        kvps << [key: it.key, value: it.value]
+                    }
+                }
+                results[it.rawScientificName] = kvps
```

Comment:
this should be a push to arrray. otherwise, i think if the two list has same scientific name it will get overwritten